### PR TITLE
Add a remote desktop protocol (RDP) scanner

### PR DESCRIPTION
**Pull request by jhart-r7**
This adds a simple RDP scanner to metasploit.  It sends the first in a series of packets necessary to establish an RDP connection and simply confirms whether the target endpoint is RDP or not.   

## Verification

- [x] Start `msfconsole`
- [x] `use modules/auxiliary/scanner/rdp/rdp_scanner.rb`
- [x] `set RHOSTS <some hosts without RDP>`
- [x] `run`
- [x] **Verify** that no hosts have RDP identified
- [x] `set RHOSTS <some hosts with RDP>`
- [x] `run`
- [x] **Verify** that all hosts have RDP identified
- [x] **Document** the thing and how it works

### CODE REVIEW

**Comment by bwatters-r7** [comment]
## Windows 10 test Range Results:
Inside the range were 12 Windows 10 machines, two each of
Win 10x86 Build 1511
Win 10x86 Build 1607
Win 10x86 Build 1703
Win 10x64 Build 1511
Win 10x64 Build 1607
Win 10x64 Build 1703
Each pair was identical except that one had RDP enabled and one did not.
TL;DR: I expect six matches.
```
msf > use auxiliary/scanner/rdp/
use auxiliary/scanner/rdp/ms12_020_check  use auxiliary/scanner/rdp/rdp_scanner     
msf > use auxiliary/scanner/rdp/rdp_scanner 
msf auxiliary(rdp_scanner) > show options

Module options (auxiliary/scanner/rdp/rdp_scanner):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   CredSSP    true             yes       Whether or not to request CredSSP
   EarlyUser  false            yes       Whether to support Earlier User Authorization Result PDU
   RHOSTS                      yes       The target address range or CIDR identifier
   RPORT      3389             yes       The target port (TCP)
   THREADS    1                yes       The number of concurrent threads
   TLS        true             yes       Wheter or not request TLS security

msf auxiliary(rdp_scanner) > set rhosts 192.168.134.0/24
rhosts => 192.168.134.0/24
msf auxiliary(rdp_scanner) > run

[*] Scanned  26 of 256 hosts (10% complete)
[*] Scanned  52 of 256 hosts (20% complete)
[*] Scanned  77 of 256 hosts (30% complete)
[*] Scanned 103 of 256 hosts (40% complete)
[*] Scanned 128 of 256 hosts (50% complete)
[*] Scanned 154 of 256 hosts (60% complete)
[+] 192.168.134.160:3389  - Identified RDP
[+] 192.168.134.161:3389  - Identified RDP
[+] 192.168.134.162:3389  - Identified RDP
[+] 192.168.134.163:3389  - Identified RDP
[+] 192.168.134.165:3389  - Identified RDP
[*] Scanned 180 of 256 hosts (70% complete)
[+] 192.168.134.189:3389  - Identified RDP
[*] Scanned 205 of 256 hosts (80% complete)
[*] Scanned 231 of 256 hosts (90% complete)

```
Success

File: modules/auxiliary/scanner/rdp/rdp_scanner.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      return unless rdp?
+    rescue Rex::AddressInUse, Rex::HostUnreachable, Rex::ConnectionTimeout, Rex::ConnectionRefused, \
+           ::Errno::ETIMEDOUT, ::Timeout::Error, ::EOFError => e
+      vprint_error("error while connecting and negotiating RDP: #{e}")
+      return
+    ensure
+      disconnect
+    end
+
+    service = report_service(
```
Comment (bwatters-r7) [inline modules/auxiliary/scanner/rdp/rdp_scanner.rb]: Did you mean to do anything with `service` after the assignment?

**Comment by jhart-r7** [comment]
@bwatters-r7 one additional test that might be helpful is by disabling `TLS`, `CredSSP` and `EarlyUser`.  It is my understanding that Windows 10 will no longer respond in such a way that indicates that it is open -- I can't recall if it just RSTs the connection or if we get some error response back.  IMO it is worth testing as that is sort of the point of the options, to tailor the scan to the particulars of the configuration you are hoping to identify, or be as compatible as possible by default.

**Comment by bcoles** [comment]
It's probably outside the scope of this PR, but it would be awesome if this module scraped usernames from systems which did not make use of "Do not display last username" configuration.

**Comment by bwatters-r7** [comment]
@bcoles, that would be cool, but you're right, it would probably be better to add it later.  I was kind of surprised that we did not already have an RDP scanner in metasploit, so I was planning on landing this ASAP.
@jhart-r7 thanks for the quick turnaround, and you're right; I should have exercised those options a bit more.....

**Comment by bwatters-r7** [comment]
# Testing 45f81f3c98ba8779a327c19aabe88a2344186224
## Win 10
```
msf > use auxiliary/scanner/rdp/rdp_scanner 
msf auxiliary(rdp_scanner) > set rhosts 192.168.134.0/24
rhosts => 192.168.134.0/24
msf auxiliary(rdp_scanner) > show options

Module options (auxiliary/scanner/rdp/rdp_scanner):

   Name       Current Setting   Required  Description
   ----       ---------------   --------  -----------
   CredSSP    true              yes       Whether or not to request CredSSP
   EarlyUser  false             yes       Whether to support Earlier User Authorization Result PDU
   RHOSTS     192.168.134.0/24  yes       The target address range or CIDR identifier
   RPORT      3389              yes       The target port (TCP)
   THREADS    1                 yes       The number of concurrent threads
   TLS        true              yes       Wheter or not request TLS security

msf auxiliary(rdp_scanner) > run

[*] Scanned  26 of 256 hosts (10% complete)
[*] Scanned  52 of 256 hosts (20% complete)
[*] Scanned  77 of 256 hosts (30% complete)
[*] Scanned 103 of 256 hosts (40% complete)
[*] Scanned 128 of 256 hosts (50% complete)
[*] Scanned 154 of 256 hosts (60% complete)
[+] 192.168.134.163:3389  - Identified RDP
[+] 192.168.134.165:3389  - Identified RDP
[*] Scanned 180 of 256 hosts (70% complete)
[+] 192.168.134.189:3389  - Identified RDP
[+] 192.168.134.197:3389  - Identified RDP
[+] 192.168.134.198:3389  - Identified RDP
[+] 192.168.134.200:3389  - Identified RDP
[*] Scanned 205 of 256 hosts (80% complete)
[*] Scanned 231 of 256 hosts (90% complete)
[*] Scanned 256 of 256 hosts (100% complete)
[*] Auxiliary module execution completed


```

## Win 10 CredSSP and TLS disabled

```
msf > use auxiliary/scanner/rdp/rdp_scanner 
msf auxiliary(rdp_scanner) > set rhosts 192.168.134.0/24
rhosts => 192.168.134.0/24
msf auxiliary(rdp_scanner) > show options

Module options (auxiliary/scanner/rdp/rdp_scanner):

   Name       Current Setting   Required  Description
   ----       ---------------   --------  -----------
   CredSSP    true              yes       Whether or not to request CredSSP
   EarlyUser  false             yes       Whether to support Earlier User Authorization Result PDU
   RHOSTS     192.168.134.0/24  yes       The target address range or CIDR identifier
   RPORT      3389              yes       The target port (TCP)
   THREADS    1                 yes       The number of concurrent threads
   TLS        true              yes       Wheter or not request TLS security

msf auxiliary(rdp_scanner) > run

[*] Scanned  26 of 256 hosts (10% complete)
[*] Scanned  52 of 256 hosts (20% complete)
[*] Scanned  77 of 256 hosts (30% complete)
[*] Scanned 103 of 256 hosts (40% complete)
[*] Scanned 128 of 256 hosts (50% complete)
[*] Scanned 154 of 256 hosts (60% complete)
[+] 192.168.134.163:3389  - Identified RDP
[+] 192.168.134.165:3389  - Identified RDP
[*] Scanned 180 of 256 hosts (70% complete)
[+] 192.168.134.189:3389  - Identified RDP
[+] 192.168.134.197:3389  - Identified RDP
[+] 192.168.134.198:3389  - Identified RDP
[+] 192.168.134.200:3389  - Identified RDP
[*] Scanned 205 of 256 hosts (80% complete)

```
## Win7x64 SP0
### Remote Access with NLA Enabled
```
msf > use auxiliary/scanner/rdp/rdp_scanner 
msf auxiliary(rdp_scanner) > set rhosts 192.168.134.133
rhosts => 192.168.134.133
msf auxiliary(rdp_scanner) > run

[+] 192.168.134.133:3389  - Identified RDP
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(rdp_scanner) > set TLS false
TLS => false
msf auxiliary(rdp_scanner) > run

[+] 192.168.134.133:3389  - Identified RDP
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(rdp_scanner) > set CredSSP false
CredSSP => false
msf auxiliary(rdp_scanner) > run
```
### Changed Remote Access not to require NLA requirement
```
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(rdp_scanner) > run

[+] 192.168.134.133:3389  - Identified RDP
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(rdp_scanner) > 
```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a scanner module to enumerate RDP services with varying levels of authentication support.

**Comment by jhart-r7** [comment]
Thanks for the testing and feedback, @bwatters-r7 and @bcoles.  There is definitely a lot of room for future improvement with regards to RDP coverage in Metasploit.  This is just a start :)

File: modules/auxiliary/scanner/rdp/rdp_scanner.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'           => 'Identify endpoints speaking the Remote Desktop Protocol (RDP)',
+        'Description'    => %q(
+          This module attempts to connect to the specified Remote Desktop Protocol port
+          and determines if it speaks RDP.
```
Comment (egypt) [inline modules/auxiliary/scanner/rdp/rdp_scanner.rb]: Can you add something here (or possibly in the doc file) about why you would turn on `CredSSP` and `EarlyUser`?
